In [1]:
import numpy as np
import pandas as pd
from matplotlib import pyplot as plt

from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import mean_squared_error, accuracy_score

In [2]:
df = pd.read_csv("./car.csv", header=None)
df.head()

0      1  2  3      4     5      6
0  vhigh  vhigh  2  2  small   low  unacc
1  vhigh  vhigh  2  2  small   med  unacc
2  vhigh  vhigh  2  2  small  high  unacc
3  vhigh  vhigh  2  2    med   low  unacc
4  vhigh  vhigh  2  2    med   med  unacc

In [3]:
df.columns = ["buying", "maintain", "doors", "seats", "cargo", "safety", "worth"]
df.head()

buying maintain doors seats  cargo safety  worth
0  vhigh    vhigh     2     2  small    low  unacc
1  vhigh    vhigh     2     2  small    med  unacc
2  vhigh    vhigh     2     2  small   high  unacc
3  vhigh    vhigh     2     2    med    low  unacc
4  vhigh    vhigh     2     2    med    med  unacc

In [4]:
df.describe()

buying maintain doors seats  cargo safety  worth
count    1728     1728  1728  1728   1728   1728   1728
unique      4        4     4     3      3      3      4
top     vhigh    vhigh     2     2  small    low  unacc
freq      432      432   432   576    576    576   1210

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1728 entries, 0 to 1727
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   buying    1728 non-null   object
 1   maintain  1728 non-null   object
 2   doors     1728 non-null   object
 3   seats     1728 non-null   object
 4   cargo     1728 non-null   object
 5   safety    1728 non-null   object
 6   worth     1728 non-null   object
dtypes: object(7)
memory usage: 94.6+ KB


In [8]:
for i in df.columns:
    print(i, df[i].unique())

buying ['vhigh' 'high' 'med' 'low']
maintain ['vhigh' 'high' 'med' 'low']
doors ['2' '3' '4' '5more']
seats ['2' '4' 'more']
cargo ['small' 'med' 'big']
safety ['low' 'med' 'high']
worth ['unacc' 'acc' 'vgood' 'good']


=> Data Description: There are 1728 datapoints. None of those contain null data

The datatype of all columns are object. Proceed to encode to int

In [9]:
import category_encoders
encoder = category_encoders.OrdinalEncoder(cols=["buying", "maintain", "doors", "seats", "cargo", "safety", "worth"])
df = encoder.fit_transform(df)

/miniconda3/lib/python3.9/site-packages/statsmodels/compat/pandas.py:61: FutureWarning: pandas.Int64Index is deprecated and will be removed from pandas in a future version. Use pandas.Index with the appropriate dtype instead.
  from pandas import Int64Index as NumericIndex
/miniconda3/lib/python3.9/site-packages/category_encoders/utils.py:21: FutureWarning: is_categorical is deprecated and will be removed in a future version. Use is_categorical_dtype instead.
  elif pd.api.types.is_categorical(cols):


In [10]:
print(df.dtypes)

buying      int64
maintain    int64
doors       int64
seats       int64
cargo       int64
safety      int64
worth       int64
dtype: object


In [11]:
for i in df.columns:
    print(i, df[i].unique())

buying [1 2 3 4]
maintain [1 2 3 4]
doors [1 2 3 4]
seats [1 2 3]
cargo [1 2 3]
safety [1 2 3]
worth [1 2 3 4]


In [12]:
df.head(20)

buying  maintain  doors  seats  cargo  safety  worth
0        1         1      1      1      1       1      1
1        1         1      1      1      1       2      1
2        1         1      1      1      1       3      1
3        1         1      1      1      2       1      1
4        1         1      1      1      2       2      1
5        1         1      1      1      2       3      1
6        1         1      1      1      3       1      1
7        1         1      1      1      3       2      1
8        1         1      1      1      3       3      1
9        1         1      1      2      1       1      1
10       1         1      1      2      1       2      1
11       1         1      1      2      1       3      1
12       1         1      1      2      2       1      1
13       1         1      1      2      2       2      1
14       1         1      1      2      2       3      1
15       1         1      1      2      3       1      1
16       1         1      1      2      3       2      1
17       1         1      1      2      3       3      1
18       1         1      1      3      1       1      1
19       1         1      1      3      1       2      1

In [13]:
print(df.loc[[1528]])

      buying  maintain  doors  seats  cargo  safety  worth
1528       4         3      1      2      3       2      4


Prepare features and labels

In [14]:
X = df.drop(["worth"], axis=1)
y = df["worth"]
print(X.shape)
print(y.shape)

(1728, 6)
(1728,)


Train - Validate split

In [15]:
X_train, X_val, y_train, y_val = train_test_split(X, y, test_size=0.2, random_state=42)
print(X_train.shape)
print(y_train.shape)
print(X_val.shape)
print(y_val.shape)

(1382, 6)
(1382,)
(346, 6)
(346,)


Train Logistic Regression model

In [16]:
logi_regr = LogisticRegression()

In [17]:
logi_regr.fit(X_train, y_train)

y_pred_train = logi_regr.predict(X_train)
acc_train = accuracy_score(y_train, y_pred_train)
tr_error = mean_squared_error(y_train, y_pred_train)

y_pred_val = logi_regr.predict(X_val)
acc_val = accuracy_score(y_val, y_pred_val)
val_error = mean_squared_error(y_val, y_pred_val)

print("training accuracy: ", acc_train)
print("validation accuracy: ", acc_val)

print("training mean squared error: ", tr_error)
print("validation mean squared error: ", val_error)

training accuracy:  0.8299565846599132
validation accuracy:  0.815028901734104
training mean squared error:  0.3473227206946454
validation mean squared error:  0.29190751445086704


/miniconda3/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
